modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/BulkCopyEntriesMVCActionCommand.java

```diff
@@ -0,0 +1,194 @@
+			actionRequest, "selectedEntries");
+		long sourceFolderId = ParamUtil.getLong(
+			actionRequest, "sourceFolderId");
+
+		ServiceContext dlFileEntryServiceContext =
+			ServiceContextFactory.getInstance(
+				DLFileEntry.class.getName(), actionRequest);
+
+		ServiceContext dlFileShortcutServiceContext =
+			ServiceContextFactory.getInstance(
+				DLFileShortcut.class.getName(), actionRequest);
+
+		ServiceContext dlFolderServiceContext =
+			ServiceContextFactory.getInstance(
+				DLFolder.class.getName(), actionRequest);
+
+		try {
+			_copyEntities(
+				entityIds, destinationFolderId, destinationRepositoryId,
+				sourceFolderId, dlFileEntryServiceContext,
+				dlFileShortcutServiceContext, dlFolderServiceContext);
+		}
+		catch (PortalException portalException) {
+			String errorMessage = StringBundler.concat(
+				portalException.getMessage(), StringPool.SPACE,
+				themeDisplay.translate(
+					"documents-folders-could-not-be-copied"));
+
+			if (_errorsCount < 10) {
+				errorMessage = StringBundler.concat(
+					portalException.getMessage(), StringPool.SPACE,
+					themeDisplay.translate(
+						"the-following-documents-folders-could-not-be-copied"));
+			}
+
+			JSONPortletResponseUtil.writeJSON(
+				actionRequest, actionResponse,
+				JSONUtil.put("errorMessage", errorMessage));
+
+			hideDefaultSuccessMessage(actionRequest);
+		}
+	}
+
+	private void _checkDestinationGroup(Group group) throws PortalException {
+		if ((group != null) && group.isStaged() && !group.isStagingGroup()) {
+			throw new PortalException(
+				"cannot-copy-into-the-live-version-of-a-group");
+		}
+	}
+
+	private void _copyEntities(
+			long[] entityIds, long destinationFolderId,
+			long destinationRepositoryId, long sourceFolderId,
+			ServiceContext dlFileEntryServiceContext,
+			ServiceContext dlFileShortcutServiceContext,
+			ServiceContext dlFolderServiceContext)
+		throws PortalException {
+
+		Group group = _groupLocalService.fetchGroup(destinationRepositoryId);
+
+		_checkDestinationGroup(group);
+
+		long[] currentAndAncestorSiteAndDepotGroupIds =
+			_siteConnectedGroupGroupProvider.
+				getCurrentAndAncestorSiteAndDepotGroupIds(group.getGroupId());
+
+		for (long entityId : entityIds) {
+			DLFileShortcut dlFileShortcut =
+				_dlFileShortcutLocalService.fetchDLFileShortcut(entityId);
+
+			try {
+				if (_dlFileEntryLocalService.fetchDLFileEntry(entityId) !=
+						null) {
+
+					_dlAppService.copyFileEntry(
+						entityId, destinationFolderId, destinationRepositoryId,
+						currentAndAncestorSiteAndDepotGroupIds,
+						dlFileEntryServiceContext);
+				}
+				else if (dlFileShortcut != null) {
+					_dlAppService.copyFileShortcut(
+						entityId, destinationFolderId, destinationRepositoryId,
+						dlFileShortcutServiceContext);
+				}
+				else if (_dlFolderLocalService.fetchDLFolder(entityId) !=
+							null) {
+
+					_dlAppService.copyFolder(
+						entityId, sourceFolderId, destinationRepositoryId,
```

Comment:
entity is the repository id so maybe is 
_dlAppService.copyFolder(sourceRepositoryId, entityId...